src/tools/rbd_mirror/InstanceWatcher.cc

```diff
@@ -46,28 +51,121 @@ struct C_GetInstances : public Context {
+      on_finish(on_finish) {
   }
 
   void send() {
+    dout(20) << "C_RemoveInstanceRequest: " << this << " " << __func__ << dendl;
+
     instance_watcher.remove(this);
   }
 
   void finish(int r) override {
+    dout(20) << "C_RemoveInstanceRequest: " << this << " " << __func__ << ": r="
+             << r << dendl;
     assert(r == 0);
 
     on_finish->complete(r);
   }
 };
 
+struct C_NotifyInstanceRequest : public Context {
+  librbd::watcher::Notifier notifier;
+  std::string instance_id;
+  uint64_t request_id;
+  bufferlist bl;
+  AsyncOpTracker &op_tracker;
+  Context *on_finish;
+  librbd::watcher::NotifyResponse response;
+
+  C_NotifyInstanceRequest(librados::IoCtx &io_ctx, ContextWQ *work_queue,
+                          const std::string &instance_id, uint64_t request_id,
+                          bufferlist &&bl, AsyncOpTracker &op_tracker,
+                          Context *on_finish)
+    : notifier(work_queue, io_ctx, RBD_MIRROR_INSTANCE_PREFIX + instance_id),
+      instance_id(instance_id), request_id(request_id), bl(bl),
+      op_tracker(op_tracker), on_finish(on_finish) {
+    op_tracker.start_op();
+  }
+
+  void send() {
+    dout(20) << "C_NotifyInstanceRequest: " << this << " " << __func__ << dendl;
+
+    notifier.notify(bl, &response, this);
+  }
+
+  void finish(int r) override {
+    dout(20) << "C_NotifyInstanceRequest: " << this << " " << __func__ << ": r="
+             << r << dendl;
+
+    if (r == 0 || r == -ETIMEDOUT) {
+      bool found = false;
+      for (auto &it : response.acks) {
+        auto &bl = it.second;
+        if (it.second.length() == 0) {
+          dout(20) << "C_NotifyInstanceRequest: " << this << " " << __func__
+                   << ": no payload in ack, ignoring" << dendl;
+          continue;
+        }
+        try {
+          auto iter = bl.begin();
+          NotifyAckPayload ack;
+          ::decode(ack, iter);
+          if (ack.instance_id != instance_id) {
+            derr << "C_NotifyInstanceRequest: " << this << " " << __func__
+                 << ": ack instance_id (" << ack.instance_id << ") "
+                 << "does not match, ignoring" << dendl;
+            continue;
+          }
+          if (ack.request_id != request_id) {
+            derr << "C_NotifyInstanceRequest: " << this << " " << __func__
+                 << ": ack request_id (" << ack.request_id << ") "
+                 << "does not match, ignoring" << dendl;
+            continue;
+          }
+          r = ack.ret_val;
+          found = true;
+          break;
+        } catch (const buffer::error &err) {
+          derr << "C_NotifyInstanceRequest: " << this << " " << __func__
+               << ": failed to decode ack: " << err.what() << dendl;
+          continue;
+        }
+      }
+
+      if (!found) {
+        if (r == -ETIMEDOUT) {
+          derr << "C_NotifyInstanceRequest: " << this << " " << __func__
+               << ": resending after timeout" << dendl;
+          send();
```

Comment:
We are going to need some way to cancel in-flight notifications (e.g. leader died so we will never get a response from it). All in-flight notifications would be canceled on shut down and notifications to a specific peer would be canceled when the leader blacklists a peer due to missed heartbeats.